base_layer/core/src/base_node/comms_interface/inbound_handlers.rs

```diff
@@ -113,16 +123,34 @@ where T: BlockchainBackend
                 )
                 .await?,
             )),
-            NodeCommsRequest::GetNewBlock =>
-            // TODO: query blockchain_db and mempool to construct a new mineable block
-            {
-                unimplemented!()
+            NodeCommsRequest::GetNewBlock => {
+                let metadata = async_db::get_metadata(self.blockchain_db.clone()).await?;
+                let best_block_hash = metadata.best_block.ok_or(CommsInterfaceError::UnexpectedApiResponse)?;
+                let best_block_header =
+                    async_db::fetch_header_with_block_hash(self.blockchain_db.clone(), best_block_hash).await?;
+                let mut header = BlockHeader::from_previous(&best_block_header);
+                header.total_difficulty = best_block_header.total_difficulty + Difficulty::from(1); // TODO: the new blocks difficulty should be calculated using lwma_diff
+
```

Comment:
Currently the add_block function of blockchain_db updates the MMR roots in the block header. I think it already has to be added to the new block header here, otherwise the PoW will change when the mmr roots get updated.

Reply:
Yeah, having looked at the code:
`pub fn add_new_block(&self, block: Block) -> Result<BlockAddResult, ChainStorageError>` inkchain_database.rs  breaks our PoW and makes the block malleable.  That needs to change asap.

Reply:
BTW `add_block` does not change the roots, only `add_new_block` does, and this is only used in tests (we should slap a `cfg(test)` attribute on it to make that clear.

The miner is going to add a coinbase utxo anyway, so the MMR roots are unknown. This is kind of why I'd prefer the miner to construct the block and just get a set of transactions from the base node's mempool.

Perhaps the road forward here is to define a new type `NewBlockTemplate` that fills in the header fields that definitely won't change and provides a list of txs in the body? It won't have MMR roots or PoW but we know that it won't validate because it's not a full block

The request would then become `NodeCommsRequest::GetNewBlockTemplate`

Reply:
But how is the MMR roots going to be filled in?

The base node does not have the coinbase UTXO. 
The Miner does not have the chain state. 

Neither has the complete info to construct a block?


Perhaps we could make this a two step process. Miner requests the basenode for a newblock template. With all the info the basenode can give. The miner then constructs the coinbase utxo. Gives the block back to the basenode for the MMR root calculation. It is then given back to the miner for mining?

Reply:
Good point.

It could still be one step though if the miner supplies a coinbase UTXO with the request for a new block

Reply:
You are right CjS77, I now see the difference between `add_block` and `add_new_block`. I added the `cfg(test)` attribute to the add_new_block function. I change the request and response to `NodeCommsRequest::GetNewBlockTemplate` and `NodeCommsResponse::NewBlockTemplate`. Currently they still return a` Block`, but I will make a ticket to replace the `Block` with a new type `NewBlockTemplate`.
Constructing the coinbase UTXO requires fees from the transactions that only the Base node knows at this point.

Reply:
The miner cannot supply the basenode with a coinbase utxo. The coinbase utxo is constructed from the fees and the emission of the block. The miner doest know the fees of the block before hand.

Reply:
🤦‍♂

Reply:
Yeah, I think thats the way we have to go.

Reply:
I created a ticket to upgrade the current system with the two stage mineable block creation process.

Reply:
👍